app/src/main/java/org/mozilla/fenix/gleanplumb/MessagingFeature.kt

```diff
@@ -0,0 +1,33 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.gleanplumb
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.map
+import mozilla.components.lib.state.helpers.AbstractBinding
+import mozilla.components.support.ktx.kotlinx.coroutines.flow.ifChanged
+import org.mozilla.fenix.components.AppStore
+import org.mozilla.fenix.components.appstate.AppAction
+import org.mozilla.fenix.components.appstate.AppState
+
+class MessagingFeature(
+    val store: AppStore,
+    private val onMessageChange: (Message?) -> Unit
+) : AbstractBinding<AppState>(store) {
+
+    override suspend fun onState(flow: Flow<AppState>) {
+        flow.map { it }.ifChanged {
+            it.messagingState.messageToShow
+        }.collect { state ->
+            onMessageChange(state.messagingState.messageToShow)
+        }
+    }
+
+    override fun start() {
+        super.start()
+        store.dispatch(AppAction.MessagingAction.Evaluate)
```

Comment:
This is a good place to have it for now. I wonder if this is soon enough for you though? Maybe earlier in the Activity/Fragment lifecycle so by the time we observe, a message will be ready?

A minor optimization you could do here is fire the action _before_ you call super (which starts observing) for the same reason above.

```suggestion
        store.dispatch(AppAction.MessagingAction.Evaluate)
        super.start()
```

Reply:
Thanks for noticing this. I'll add the suggestion. I'm not completely sure how time consuming could the evaluations be. I guess ideally we would like to make the evaluation in bg thread, and post the results async. cc  @jhugman Do you have an of the performance hit of the evaluations?

Reply:
The background thread idea is a separate idea aside from this - it's more about, don't you want to start the evaluation as soon as possible (regardless of which thread it'll be on)?

Reply:
Yes 👍🏽